totalRP3/Core/Vanilla/CompanionUtilOverrides.lua

```diff
@@ -0,0 +1,322 @@
+end
+
+function CompanionDataLoaderMixin:RequestLoadSpellData(spellID, callbackFunction)
+	self:AddLoadCallback(self.pendingSpells, spellID, callbackFunction);
+	C_Spell.RequestLoadSpellData(spellID);
+end
+
+function CompanionDataLoaderMixin:OnEvent(event, ...)
+	if event == "ITEM_DATA_LOAD_RESULT" then
+		local itemID, success = ...;
+		self:InvokeLoadCallbacks(self.pendingItems, itemID, success);
+	elseif event == "SPELL_DATA_LOAD_RESULT" then
+		local spellID, success = ...;
+		self:InvokeLoadCallbacks(self.pendingSpells, spellID, success);
+	end
+end
+
+local CompanionDataLoader = CreateFrame("Frame")
+FrameUtil.SpecializeFrameWithMixins(CompanionDataLoader, CompanionDataLoaderMixin);
+
+local function RegisterCompanionPet(petInfo)
+	staticCompanionPetsByID[petInfo.id] = petInfo;
+	staticCompanionPetsByCreatureID[petInfo.creatureID] = petInfo;
+	staticCompanionPetsBySpellID[petInfo.spellID] = petInfo;
+end
+
+local function RegisterCompanionMount(mountInfo)
+	staticMountsByID[mountInfo.mountID] = mountInfo;
+end
+
+do
+	for _, petInfo in ipairs(TRP3_CompanionPetData) do
+		local function OnPetItemLoaded(itemID, success)
+			if not success then
+				return;
+			end
+
+			petInfo = CreateFromMixins(petInfo);
+			petInfo.id = tostring(petInfo.speciesID);
+			petInfo.name = C_Item.GetItemNameByID(itemID);
+			petInfo.icon = C_Item.GetItemIconByID(itemID);
+
+			RegisterCompanionPet(petInfo);
+		end
+
+		CompanionDataLoader:RequestLoadItemData(petInfo.itemID, OnPetItemLoaded);
+	end
+
+	local playerFaction = UnitFactionGroup("player");
+
+	for _, mountInfo in ipairs(TRP3_CompanionMountData) do
+		local function OnMountSpellLoaded(spellID, success)
+			if not success then
+				return;
+			end
+
+			mountInfo = CreateFromMixins(mountInfo);
+			mountInfo.name = C_Spell.GetSpellName(spellID);
+			mountInfo.icon = C_Spell.GetSpellTexture(spellID);
+
+			RegisterCompanionMount(mountInfo);
+		end
+
+		-- Only load mounts relevant for the player faction.
+		if mountInfo.factionGroup == nil or mountInfo.factionGroup == playerFaction then
+			CompanionDataLoader:RequestLoadSpellData(mountInfo.spellID, OnMountSpellLoaded);
+		end
+	end
+end
+
+-- On client flavors without a pet journal we need to be creative with how we
+-- determine what companion pet is actually summoned.
+--
+-- Our approach is to monitor for successful spellcasts whose spell IDs
+-- are associated with that of a known companion pet. We assume that the
+-- last successful cast will represent the current battle pet.
+--
+-- Note that we can't tell when a companion pet is dismissed, so our query
+-- data will always contain the data for the last-successful cast even if
+-- that cast dismissed the pet. Realistically this should be fine since
+-- if it's dismissed other players can't see the unit to request the data
+-- anyway.
+--
+-- For persistence across UI reloads we store the summoned pet data in our
+-- saved variables, and reset it upon initial login of a new character.
```

Comment:
This logic was relocated from RegisterCompanionsMain, where it was horribly broken - we were never passing the cast GUIDs or isInitialLogin parameters through to the callbacks.